h3/src/frame/mod.rs

```diff
@@ -1,3 +1,284 @@
-#[allow(dead_code)]
-// TODO: remove once methods are effectively used through public API.
 mod buf;
+
+use std::{
+    io,
+    task::{Context, Poll},
+};
+
+use bytes::{Buf, Bytes};
+
+use crate::{
+    proto::{
+        frame::{self, Frame},
+        ErrorCode,
+    },
+    quic::RecvStream,
+};
+use buf::BufList;
+
+pub struct FrameStream<R: RecvStream> {
+    recv: R,
+    bufs: BufList<Bytes>,
+    decoder: FrameDecoder,
+    remaining_data: Option<u64>,
+}
+
+impl<R> FrameStream<R>
+where
+    R: RecvStream<Buf = Bytes>,
+{
+    pub fn poll_next(&mut self, cx: &mut Context<'_>) -> Poll<Result<Option<Frame>, Error>> {
+        assert!(self.remaining_data.is_none());
+
+        let end = self.try_recv(cx)?;
+
+        match self.decoder.decode(&mut self.bufs)? {
+            Some(Frame::Data { len }) => {
+                self.remaining_data = Some(len);
+                Poll::Ready(Ok(Some(Frame::Data { len })))
+            }
+            Some(frame) => Poll::Ready(Ok(Some(frame))),
+            None if end => Poll::Ready(Ok(None)),
+            None => Poll::Pending,
+        }
+    }
+
+    pub fn poll_data(&mut self, cx: &mut Context<'_>) -> Poll<Result<Option<Bytes>, Error>> {
+        match self.remaining_data {
+            None => return Poll::Ready(Ok(None)),
+            Some(0) => {
+                self.remaining_data = None;
+                return Poll::Ready(Ok(None));
+            }
+            _ => (),
+        }
+
+        let end = self.try_recv(cx)?;
+
+        let data = self.bufs.bytes();
+        match (data.remaining(), end) {
+            (0, true) => return Poll::Ready(Ok(None)),
+            (0, false) => return Poll::Pending,
+            (x, _) => *self.remaining_data.as_mut().unwrap() -= x as u64,
+        };
+
+        Poll::Ready(Ok(Some(Bytes::from(data))))
```

Comment:
@seanmonstar How can I achieve this ? The borrow checker says it needs `'static`.

Reply:
To copy into a new `Bytes`, you'd need `Bytes::copy_from_slice(data)`. I think that's perfectly fine for now, let's get it working, and we can see about less copies in a separate PR.

Reply:
I quite unhappy with the solution I just came up with, but it should work.